### PR TITLE
Empty head block's data when the block is reset

### LINKED ISSUE #884: Deque leaks memory
Deque uses `data::Vector{T}` in `DequeBlock{T}`, but it only empties this vector in rear blocks, never touching this vector in the head block, instead it only invokes `reset!` on it. But `reset!` doesn't unset index on data elements that are no longer valid in this head block, causing these objects not to be garbage collected.

I verified this by taking heap snapshot after inserting and dequeuing all elements and running `GC.gc(true)` - the elements were still in the heap.

Not sure if other data structures have this behaviour, I only checked for Deque (and by proxy Queue)

### PR DESCRIPTION
**Pull request by eulerkochy**
Fixes #884

### CODE REVIEW

File: src/deque.jl [hunk 1]
```diff
@@ -38,6 +38,8 @@ isrear(blk::DequeBlock) =  blk.next === blk
 # reset the block to empty, and position
 
 function reset!(blk::DequeBlock{T}, front::Int) where T
+    empty!(blk.data)
+    blk.data = Vector{T}(undef, blk.capa)
```
Comment (oxinabox) [inline src/deque.jl:42]: Why empty it if its just going to be over written with a newly allocated array?
I would think just:
```suggestion
    empty!(blk.data)
```
would be enough.
(and it's nonallocating)

Comment (eulerkochy) [reply src/deque.jl:42]: The underlying the assumption of a block is that it's data field will have already `data` field allocated with `capa` number of elements. Just emptying and not allocating block's `data` leads to the problem that once you have cleared the deque, you can't push elements into it again. That's why I've added new testset `push! after empty!` to check this scenario.

Comment (oxinabox) [reply src/deque.jl:42]: So we want:
```suggestion
    # Fill it with undefs back to capacity
    empty!(blk.data); resize!(blk.data, blk.capa)
```
?

Comment (eulerkochy) [reply src/deque.jl]: yes. made this change

**Review by oxinabox — APPROVED** [review]

**Pull request merged**